### PR TITLE
Error handling simplification in the kit

### PR DESCRIPTION
**Pull request by StKostyantin**
This pull request for simplification mechanism for handling errors in the kit. Also, it fixes the #871 issue.

### CODE REVIEW

File: config/stripe/subscription.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 export default {
-  enabled: true,
```
Comment (evgenijolshanskij) [inline config/stripe/subscription.js]: How these changes relate to error handling?

File: packages/mobile/.spinrc.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ const config = {
       dllBuildDir: 'build/android/dll',
       stack: ['react-native', 'android'],
       defines: {
-        __CLIENT__: true
+        __CLIENT__: false
```
Comment (evgenijolshanskij) [inline packages/mobile/.spinrc.js]: Should not be changed.

File: modules/contact/server-ts/resolvers.ts [hunk 3]
```diff
@@ -28,10 +32,8 @@ export default () => ({
         });
       } catch (e) {
         log.error(e);
-        return { errors: [{ field: 'serverError', message: t('contact:sendError') }] };
+        throw new ApolloError(t('contact:sendError'), 'CONTACT_DENIED');
```
Comment (evgenijolshanskij) [inline modules/contact/server-ts/resolvers.ts]: Why did you choose `ApolloError` class while the library provides more specific `UserInputError`? (https://www.apollographql.com/docs/apollo-server/features/errors.html#Augmenting-error-details)

File: config/stripe/subscription.js (hunk 1 above)
Comment (StKostyantin) [reply config/stripe/subscription.js]: Module payment has been changed within this issue and to start this module I should have enabled this module, and after making changes disable.

File: packages/mobile/.spinrc.js (hunk 2 above)
Comment (StKostyantin) [reply packages/mobile/.spinrc.js]: fix

File: modules/contact/server-ts/resolvers.ts [hunk 4]
```diff
@@ -13,10 +15,12 @@ interface ContactInput {
...
     async contact(obj: any, { input }: ContactInput, { mailer, req: { t } }: any) {
-      const errors = new FieldError(validate(input, contactFormSchema));
+      // TODO if import ApolloError that get error 'app is not defined'
+      const { ApolloError, UserInputError } = require('apollo-server-express');
 
-      if (errors.hasAny()) {
-        return { errors: errors.getErrors() };
+      const errors = validate(input, contactFormSchema);
+      if (!isEmpty(errors)) {
+        throw new UserInputError('Failed to get events due to validation errors', { errors });
```
Comment (evgenijolshanskij) [inline modules/contact/server-ts/resolvers.ts]: All the texts coming from the server should be translated.

File: config/stripe/subscription.js (hunk 1 above)
Comment (StKostyantin) [reply config/stripe/subscription.js]: fix

File: modules/contact/server-ts/resolvers.ts (hunk 4 above)
Comment (StKostyantin) [reply modules/contact/server-ts/resolvers.ts]: Add translated for error

File: modules/contact/server-ts/__tests__/Contact.spec.ts [hunk 5]
```diff
@@ -12,19 +12,19 @@ describe('Contact API works', () => {
...
         }
-      }
-    });
-
-    expect(data.contact)
-      .to.have.property('errors')
-      .with.length(3);
+      });
+    } catch (e) {
+      expect(e);
```
Comment (Werter12) [inline modules/contact/server-ts/__tests__/Contact.spec.ts]: I don't think this assertion is right.

Comment (StKostyantin) [reply modules/contact/server-ts/__tests__/Contact.spec.ts]: fix

File: modules/payments/client-react/stripe/subscription/containers/UpdateCreditCard.tsx [hunk 6]
```diff
@@ -36,29 +36,23 @@ class UpdateCreditCard extends React.Component<UpdateCreditCardProps, { [key: st
 
     try {
       // create credit card token
-      try {
-        preparedCreditCard = await createCreditCardToken(creditCardInput, stripe);
-        if (preparedCreditCard.error) {
-          this.setState({ submitting: false, error: t('stripeError') });
```
Comment (alekseyzadorozhniy) [inline modules/payments/client-react/stripe/subscription/containers/UpdateCreditCard.tsx]: Looks like you forgot to remove "error" field from the component state and from the render function

File: modules/payments/client-react/stripe/subscription/containers/AddSubscription.tsx [hunk 7]
```diff
@@ -24,8 +24,7 @@ class AddSubscription extends React.Component<AddSubscriptionProps, { [key: stri
   constructor(props: AddSubscriptionProps) {
     super(props);
     this.state = {
-      submitting: false,
-      error: null
```
Comment (alekseyzadorozhniy) [inline modules/payments/client-react/stripe/subscription/containers/AddSubscription.tsx]: Looks like you forgot to remove the use of "error" field from the render function

File: modules/payments/client-react/stripe/subscription/containers/UpdateCreditCard.tsx (hunk 6 above)
Comment (StKostyantin) [reply modules/payments/client-react/stripe/subscription/containers/UpdateCreditCard.tsx]: Fixed

File: modules/payments/client-react/stripe/subscription/containers/AddSubscription.tsx (hunk 7 above)
Comment (StKostyantin) [reply modules/payments/client-react/stripe/subscription/containers/AddSubscription.tsx]: Fixed

File: modules/user/server-ts/resolvers.js [hunk 8]
```diff
@@ -195,7 +193,7 @@ export default pubsub => ({
 
           return { user };
         } catch (e) {
-          return { errors: e };
+          console.log(e);
```
Comment (alekseyzadorozhniy) [inline modules/user/server-ts/resolvers.js]: Shouldn't you return an error here?

File: modules/user/server-ts/auth/password/resolvers.js [hunk 9]
```diff
@@ -37,74 +34,67 @@ export default () => ({
...
+        // async email
+        jwt.sign({ user: pick(user, 'id') }, settings.user.secret, { expiresIn: '1d' }, (err, emailToken) => {
+          const encodedToken = Buffer.from(emailToken).toString('base64');
+          const url = `${__WEBSITE_URL__}/confirmation/${encodedToken}`;
+          mailer.sendMail({
+            from: `${settings.app.name} <${process.env.EMAIL_USER}>`,
+            to: user.email,
+            subject: 'Confirm Email',
+            html: `<p>Hi, ${user.username}!</p>
+            <p>Welcome to ${settings.app.name}. Please click the following link to confirm your email:</p>
```
[the remark this answers is not in the document]
Comment (StKostyantin) [reply modules/user/server-ts/auth/password/resolvers.js]: This is beyond the scope of this pull request. For this task, I will create a new pull request and make translate.

File: modules/user/server-ts/resolvers.js (hunk 8 above)
Comment (StKostyantin) [reply modules/user/server-ts/resolvers.js]: Fixed

File: modules/user/client-react/containers/UserOperations.js [hunk 10]
```diff
@@ -43,7 +43,7 @@ const withUsersDeleting = Component =>
             return { errors: deleteUser.errors };
           }
         } catch (e) {
-          console.log(e.graphQLErrors);
+          console.error(e);
```
Comment (vlasenko) [inline modules/user/client-react/containers/UserOperations.js]: We should use log.error here

Comment (StKostyantin) [reply modules/user/client-react/containers/UserOperations.js]: Fixed

File: modules/forms/client-react/FormError.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import { isApolloError } from 'apollo-client';
+import { GraphQLError } from 'graphql';
+
+export const isFormError = (err: any) => err instanceof FormError;
+
+export class FormError {
+  public _errors: { [key: string]: any };
```
Comment (evgenijolshanskij) [inline modules/forms/client-react/FormError.ts]: I guess this field is not supposed to be `public`.

File: modules/forms/client-react/FormError.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+  constructor(errorMsg: string, err?: any) {
+    if (!!err) {
```
Comment (evgenijolshanskij) [inline modules/forms/client-react/FormError.ts]: `if` statement automatically converts a value into `Boolean`, so there is no need to it explicitly.

File: modules/contact/server-ts/resolvers.ts [hunk 13]
```diff
@@ -28,10 +29,8 @@ export default () => ({
         });
       } catch (e) {
         log.error(e);
-        return { errors: [{ field: 'serverError', message: t('contact:sendError') }] };
+        throw new Error(t('contact:sendError'));
```
Comment (evgenijolshanskij) [inline modules/contact/server-ts/resolvers.ts]: Not sure about this. If we decide to jump on Apollo errors handling then we should use the `ApolloError` everywhere (https://www.apollographql.com/docs/apollo-server/features/errors.html#Other-errors). Probably it is better to create some general custom error based on `ApolloError` just like others implemented - https://github.com/apollographql/apollo-server/blob/master/packages/apollo-server-errors/src/index.ts#L169. Or let's each module provides it's own set of errors if it's needed, I don't know.

File: modules/payments/client-react/stripe/subscription/components/SubscriptionCardFormView.tsx [hunk 14]
```diff
@@ -34,19 +33,25 @@ const SubscriptionCardFormView = (props: SubscriptionCardFormViewProps) => {
...
   mapPropsToValues: () => ({ name: '' }),
-  async handleSubmit({ name }, { props }: { props: SubscriptionCardFormViewProps }) {
+  handleSubmit({ name }, { setErrors, props }) {
     const { stripe, onSubmit } = props;
-    onSubmit({ name }, stripe);
+    onSubmit({ name }, stripe).catch((e: { [key: string]: any }) => {
+      if (isFormError(e)) {
+        setErrors(e.errors);
+      } else {
+        throw e;
```
Comment (evgenijolshanskij) [inline modules/payments/client-react/stripe/subscription/components/SubscriptionCardFormView.tsx]: Let's stick to some common solution.

File: modules/validation/common-react/FieldError.ts [hunk 15]
```diff
@@ -1,7 +1,7 @@
-export type Errors = Array<{ field: string; message: string }> | { [key: string]: string };
+export type Errors = Array<{ field: string; message: string }> | { [key: string]: any };
 
 export class FieldError {
-  public errors: { [key: string]: string };
+  public errors: { [key: string]: any };
```
Comment (evgenijolshanskij) [inline modules/validation/common-react/FieldError.ts]: Broken encapsulation.

Comment (StKostyantin) [reply modules/validation/common-react/FieldError.ts]: Fixed

File: modules/forms/client-react/FormError.ts (hunk 11 above)
Comment (StKostyantin) [reply modules/forms/client-react/FormError.ts]: Fixed

File: modules/forms/client-react/FormError.ts (hunk 12 above)
Comment (StKostyantin) [reply modules/forms/client-react/FormError.ts]: Fixed

**Comment by vlasenko** [comment]
Thank you @StKostyantin ! It was a serious effort. Looks nice!

**Pull request merged**